src/cascade/input_data/emr.py

```diff
@@ -0,0 +1,219 @@
+            )
+
+            csmr_mean = mean_interp["both"].integral(
+                prevalence_measurement.age_lower,
+                prevalence_measurement.age_upper,
+                prevalence_measurement.time_lower,
+                prevalence_measurement.time_upper,
+            )
+            csmr_mean /= divisor
+
+            csmr_stderr = stderr_interp["both"].integral(
+                prevalence_measurement.age_lower,
+                prevalence_measurement.age_upper,
+                prevalence_measurement.time_lower,
+                prevalence_measurement.time_upper,
+            )
+            csmr_stderr /= divisor
+
+        emr_mean = csmr_mean / prevalence_measurement["mean"]
+
+        # Propagation of uncertainty for f = A/B gives a standard deviation of
+        # sqrt( (sigma_a)^2 / A^2 + (sigma_b)^2 / B^2 - 2 (sigma_ab) / (A B) )
+        # but we don't know the covariance sigma_ab, so we use an upper bound on the error by
+        # assuming it's zero.
+        emr_stderr = np.sqrt(
+            (prevalence_measurement.standard_error / prevalence_measurement["mean"]) ** 2
+            + (csmr_stderr / csmr_mean) ** 2
+        )
+        return pd.Series({"mean": float(emr_mean), "standard_error": float(emr_stderr)})
+
+    return emr.merge(prevalence.apply(emr_mean, "columns"), left_index=True, right_index=True)
+
+
+def add_emr_from_prevalence(model_context, execution_context):
+    r"""Estimate excess mortality from the supplied prevalence measurements and
+    cause specific mortality estimates from GBD using the formula:
+    :math:`\chi=\frac{P}{{}_nm_x^c}
+    """
+    MATHLOG.debug("Calculating excess mortality using: EMR=CSMR/prevalence")
+    prevalence = _prepare_prevalence(model_context.input_data.observations)
+    csmr = get_cause_specific_mortality_data(execution_context)
+    csmr = _prepare_csmr(execution_context, csmr)
+
+    emr = _calculate_emr_from_csmr_and_prevalence(csmr, prevalence)
+
+    model_context.input_data.observations = pd.concat([model_context.input_data.observations, emr])
+
+
+def _calculate_emr_from_csmr_and_prevalence(all_csmr, all_prevalence):
+    MATHLOG.debug(
+        dedent(
+            """
+    For every sex/location combination in the prevalence data we construct a
+    linear interpolation of CSMR over age and time and use the integral of that
+    to estimate the mean CSMR in the age/time region of each prevalence
+    measurement which is then used to calculate EMR.
+    """
+        )
+    )
+    excess_mortality_chunks = []
+    non_interpolatable_groups = all_prevalence[["sex_id", "node_id"]].drop_duplicates()
+    for (sex_id, node_id) in non_interpolatable_groups.values:
+        csmr = all_csmr.query("sex_id == @sex_id and node_id == @node_id")
+        prevalence = all_prevalence.query("sex_id == @sex_id and node_id == @node_id")
+        if csmr.empty:
+            MATHLOG.debug(
```

Comment:
Odd question. Running `python -O` turns off debug statements, right? No, it turns off `__DEBUG__` but leaves logging debug statements. nvmd.

Reply:
Yeah, it would be odd for it to effect the logger.